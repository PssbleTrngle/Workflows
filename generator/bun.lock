{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "@pssbletrngle/github-meta-generator",
      "dependencies": {
        "handlebars": "^4.7.8",
        "prettier": "^3.6.2",
        "yaml": "^2.8.1",
      },
      "devDependencies": {
        "@types/bun": "latest",
      },
      "peerDependencies": {
        "typescript": "^5",
      },
    },
  },
  "packages": {
    "@types/bun": ["@types/bun@1.3.1", "", { "dependencies": { "bun-types": "1.3.1" } }, "sha512-4jNMk2/K9YJtfqwoAa28c8wK+T7nvJFOjxI4h/7sORWcypRNxBpr+TPNaCfVWq70tLCJsqoFwcf0oI0JU/fvMQ=="],

    "@types/node": ["@types/node@24.9.1", "", { "dependencies": { "undici-types": "~7.16.0" } }, "sha512-QoiaXANRkSXK6p0Duvt56W208du4P9Uye9hWLWgGMDTEoKPhuenzNcC4vGUmrNkiOKTlIrBoyNQYNpSwfEZXSg=="],

    "@types/react": ["@types/react@19.2.2", "", { "dependencies": { "csstype": "^3.0.2" } }, "sha512-6mDvHUFSjyT2B2yeNx2nUgMxh9LtOWvkhIU3uePn2I2oyNymUAX1NIsdgviM4CH+JSrp2D2hsMvJOkxY+0wNRA=="],

    "bun-types": ["bun-types@1.3.1", "", { "dependencies": { "@types/node": "*" }, "peerDependencies": { "@types/react": "^19" } }, "sha512-NMrcy7smratanWJ2mMXdpatalovtxVggkj11bScuWuiOoXTiKIu2eVS1/7qbyI/4yHedtsn175n4Sm4JcdHLXw=="],

    "csstype": ["csstype@3.1.3", "", {}, "sha512-M1uQkMl8rQK/szD0LNhtqxIPLpimGm8sOBwU7lLnCpSbTyY3yeU1Vc7l4KT5zT4s/yOxHH5O7tIuuLOCnLADRw=="],

    "handlebars": ["handlebars@4.7.8", "", { "dependencies": { "minimist": "^1.2.5", "neo-async": "^2.6.2", "source-map": "^0.6.1", "wordwrap": "^1.0.0" }, "optionalDependencies": { "uglify-js": "^3.1.4" }, "bin": { "handlebars": "bin/handlebars" } }, "sha512-vafaFqs8MZkRrSX7sFVUdo3ap/eNiLnb4IakshzvP56X5Nr1iGKAIqdX6tMlm6HcNRIkr6AxO5jFEoJzzpT8aQ=="],

    "minimist": ["minimist@1.2.8", "", {}, "sha512-2yyAR8qBkN3YuheJanUpWC5U3bb5osDywNB8RzDVlDwDHbocAJveqqj1u8+SVD7jkWT4yvsHCpWqqWqAxb0zCA=="],

    "neo-async": ["neo-async@2.6.2", "", {}, "sha512-Yd3UES5mWCSqR+qNT93S3UoYUkqAZ9lLg8a7g9rimsWmYGK8cVToA4/sF3RrshdyV3sAGMXVUmpMYOw+dLpOuw=="],

    "prettier": ["prettier@3.6.2", "", { "bin": { "prettier": "bin/prettier.cjs" } }, "sha512-I7AIg5boAr5R0FFtJ6rCfD+LFsWHp81dolrFD8S79U9tb8Az2nGrJncnMSnys+bpQJfRUzqs9hnA81OAA3hCuQ=="],

    "source-map": ["source-map@0.6.1", "", {}, "sha512-UjgapumWlbMhkBgzT7Ykc5YXUT46F0iKu8SGXq0bcwP5dz/h0Plj6enJqjz1Zbq2l5WaqYnrVbwWOWMyF3F47g=="],

    "typescript": ["typescript@5.9.3", "", { "bin": { "tsc": "bin/tsc", "tsserver": "bin/tsserver" } }, "sha512-jl1vZzPDinLr9eUt3J/t7V6FgNEw9QjvBPdysz9KfQDD41fQrC2Y4vKQdiaUpFT4bXlb1RHhLpp8wtm6M5TgSw=="],

    "uglify-js": ["uglify-js@3.19.3", "", { "bin": { "uglifyjs": "bin/uglifyjs" } }, "sha512-v3Xu+yuwBXisp6QYTcH4UbH+xYJXqnq2m/LtQVWKWzYc1iehYnLixoQDN9FH6/j9/oybfd6W9Ghwkl8+UMKTKQ=="],

    "undici-types": ["undici-types@7.16.0", "", {}, "sha512-Zz+aZWSj8LE6zoxD+xrjh4VfkIG8Ya6LvYkZqtUQGJPZjYl53ypCaUwWqo7eI0x66KBGeRo+mlBEkMSeSZ38Nw=="],

    "wordwrap": ["wordwrap@1.0.0", "", {}, "sha512-gvVzJFlPycKc5dZN4yPkP8w7Dc37BtP1yczEneOb4uq34pXZcvrtRTmWV8W+Ume+XCxKgbjM+nevkyFPMybd4Q=="],

    "yaml": ["yaml@2.8.1", "", { "bin": { "yaml": "bin.mjs" } }, "sha512-lcYcMxX2PO9XMGvAJkJ3OsNMw+/7FKes7/hgerGUYWIoWu5j/+YQqcZr5JnPZWzOsEBgMbSbiSTn/dv/69Mkpw=="],
  }
}
