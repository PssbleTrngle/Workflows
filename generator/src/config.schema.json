{
   "$schema": "https://json-schema.org/draft/2020-12/schema",
   "$id": "https://possible-triangle.com/github-metadata-sync-config.schema.json",
   "title": "MetadataConfig",
   "description": "config file for the github metadata sync app",
   "type": "object",
   "properties": {
      "type": {
         "description": "the type of repository this is",
         "enum": [
            "minecraft"
         ]
      }
   },
   "required": [
      "type"
   ],
   "allOf": [
      {
         "if": {
            "properties": {
               "type": {
                  "const": "minecraft"
               }
            }
         },
         "then": {
            "properties": {
               "loaders": {
                  "description": "the mod loader this mod supports",
                  "type": "array",
                  "uniqueItems": true,
                  "items": {
                     "type": "string"
                  }
               },
               "versions": {
                  "description": "the minecraft versions this mod supports",
                  "type": "array",
                  "uniqueItems": true,
                  "items": {
                     "type": "string"
                  }
               }
            }
         }
      }
   ]
}