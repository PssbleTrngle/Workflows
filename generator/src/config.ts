import z from "zod";

const schema = z.object({
  type: z.enum(["minecraft"]),
});

export type ConfigSchema = z.infer<typeof schema>;

export function validateConfig(input: unknown) {
  return schema.parse(input);
}

export const configSchema = z.toJSONSchema(schema);
