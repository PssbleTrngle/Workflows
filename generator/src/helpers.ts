export default {
  gt: (a: number, b: number) => a > b,
  lt: (a: number, b: number) => a < b,
  gte: (a: number, b: number) => a >= b,
  lte: (a: number, b: number) => a <= b,
  multiple: (value: number) => value > 1,
};
