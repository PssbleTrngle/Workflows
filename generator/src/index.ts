export { configSchema, validateConfig } from "./config";
export type { ConfigSchema } from "./config";
export { generateWithConfig, type TemplateData } from "./generator";
export { generateIssueTemplate } from "./issueTemplates/generate";
export { isGenerated } from "./meta";
export { generateWorkflow } from "./workflows/generate";
