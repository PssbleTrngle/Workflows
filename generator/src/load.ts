import { compile } from "handlebars";
import { join } from "node:path";

async function loadFile(...path: string[]) {
  const file = Bun.file(join(__dirname, "templates", ...path));
  return file.text();
}

export async function compileFile(...path: string[]) {
  return compile(await loadFile(...path));
}
