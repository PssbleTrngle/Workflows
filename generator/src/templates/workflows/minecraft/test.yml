name: Test

on:
   workflow_dispatch:
    inputs:
      upload:
        description: "Upload Artifact"
        default: false
        required: true
        type: boolean

   pull_request:
      types: [opened, synchronize]
{{#unless sonar}}
   push:
      branches: ["[0-9]+.[0-9]+.[0-9x]+"]
{{/unless }}

env:
   CI: true

concurrency:
   group: {{ var 'github.workflow' }}-{{ var 'github.ref' }}
   cancel-in-progress: true

jobs:
   test:
      runs-on: ubuntu-latest
      timeout-minutes: 20
      permissions:
         contents: read
      steps:
         - name: Checkout
           uses: actions/checkout@v4
              
         {{> java }}


         - name: Build
           run: ./gradlew build check

         - uses: actions/upload-artifact@v4
           with:
              name: build
              path: "**/build/libs/*.jar"