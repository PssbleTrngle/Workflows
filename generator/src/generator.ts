import { basename, join } from "node:path";
import type { ConfigSchema } from "./config";
import { generateIssueTemplate } from "./issueTemplates/generate";
import { listTemplates } from "./load";
import { generateWorkflow } from "./workflows/generate";

export type Acceptor = (path: string, content: string) => Promise<void>;

export type TemplateData = Record<string, unknown>;

export type Generator = (path: string, data?: TemplateData) => Promise<string>;

type SchemaType = {
  key: keyof ConfigSchema;
  path: string;
  generate: Generator;
};

const schemaTypes: SchemaType[] = [
  {
    key: "issueTemplates",
    path: ".github/ISSUE_TEMPLATE",
    generate: generateIssueTemplate,
  },
  { key: "workflows", path: ".github/workflows", generate: generateWorkflow },
];

function check(config: ConfigSchema, key: string) {
  return key in config && config[key as keyof ConfigSchema] === false;
}

export async function generateWithConfig(
  config: ConfigSchema,
  acceptor: Acceptor
) {
  await Promise.all(
    schemaTypes.map(async ({ key, path, generate }) => {
      if (key in config && check(config, key)) return;

      const schemas = await listTemplates(key, config.type);

      await Promise.all(
        schemas.map(async (file) => {
          const fileName = basename(file);
          const name = fileName.substring(0, fileName.indexOf("."));
          if (check(config, name)) return;

          const output = join(path, fileName);

          const generated = await generate(file, config);
          await acceptor(output, generated);
        })
      );
    })
  );
}
