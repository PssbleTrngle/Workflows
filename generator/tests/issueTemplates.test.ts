import { describe, expect, it } from "bun:test";
import { generateIssueTemplate, type TemplateData } from "../src";

function* optionsProvider(): Generator<[string, TemplateData]> {
  yield ["without loader and versions", {}];
  yield ["without loader and single version", { versions: ["1.21.1"] }];
  yield [
    "without loader and multiple versions",
    { versions: ["1.20.1", "1.21.1"] },
  ];
  yield ["single loader and no versions", { loaders: ["neoforge"] }];
  yield [
    "multiple loaders and no versions",
    { loaders: ["neoforge", "fabric"] },
  ];
  yield [
    "multiple loaders and single version",
    { loaders: ["neoforge", "fabric"], versions: ["1.21.1"] },
  ];
  yield [
    "multiple loaders and multiple versions",
    { loaders: ["neoforge", "fabric"], versions: ["1.20.1", "1.21.1"] },
  ];
}

describe("minecraft bug templates", () => {
  for (const [key, options] of optionsProvider()) {
    it(`generates bug template ${key}`, async () => {
      const generated = await generateIssueTemplate(
        "minecraft/bug_report.yml",
        options
      );
      expect(generated).toMatchSnapshot(`bug template ${key}`);
    });
  }
});
