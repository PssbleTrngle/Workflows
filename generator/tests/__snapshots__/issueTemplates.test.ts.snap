// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`minecraft bug templates generates bug template without loader and versions: bug template without loader and versions 1`] = `
"# <meta>
#    <source>https://github.com/PssbleTrngle/Workflows</source>
#    <version>0.0.0-dev</version>
#    <timestamp>2025-01-01T00:00:00.000Z</timestamp>
# </meta>

name: Bug Report
description: File a bug report
labels: ["bug"]

body:
  - type: textarea
    id: what-happened
    attributes:
      label: What happened?
      description: Also tell us, what did you expect to happen?
      placeholder: Tell us what you see!
    validations:
      required: true

  - type: input
    id: version-version
    attributes:
      label: Mod Version
    validations:
      required: true

  - type: input
    id: logs
    attributes:
      label: log output
      description: Please copy and paste your log output to https://mclo.gs/ or a similar site and paste the link here.
"
`;

exports[`minecraft bug templates generates bug template without loader and single version: bug template without loader and single version 1`] = `
"# <meta>
#    <source>https://github.com/PssbleTrngle/Workflows</source>
#    <version>0.0.0-dev</version>
#    <timestamp>2025-01-01T00:00:00.000Z</timestamp>
# </meta>

name: Bug Report
description: File a bug report
labels: ["bug"]

body:
  - type: textarea
    id: what-happened
    attributes:
      label: What happened?
      description: Also tell us, what did you expect to happen?
      placeholder: Tell us what you see!
    validations:
      required: true

  - type: input
    id: version-version
    attributes:
      label: Mod Version
    validations:
      required: true

  - type: input
    id: logs
    attributes:
      label: log output
      description: Please copy and paste your log output to https://mclo.gs/ or a similar site and paste the link here.
"
`;

exports[`minecraft bug templates generates bug template without loader and multiple versions: bug template without loader and multiple versions 1`] = `
"# <meta>
#    <source>https://github.com/PssbleTrngle/Workflows</source>
#    <version>0.0.0-dev</version>
#    <timestamp>2025-01-01T00:00:00.000Z</timestamp>
# </meta>

name: Bug Report
description: File a bug report
labels: ["bug"]

body:
  - type: textarea
    id: what-happened
    attributes:
      label: What happened?
      description: Also tell us, what did you expect to happen?
      placeholder: Tell us what you see!
    validations:
      required: true

  - type: input
    id: version-version
    attributes:
      label: Mod Version
    validations:
      required: true

  - type: dropdown
    id: minecraft-version
    attributes:
      label: Minecraft Version
      multiple: false
      options:
        - "1.20.1"
        - "1.21.1"
    validations:
      required: true

  - type: input
    id: logs
    attributes:
      label: log output
      description: Please copy and paste your log output to https://mclo.gs/ or a similar site and paste the link here.
"
`;

exports[`minecraft bug templates generates bug template single loader and no versions: bug template single loader and no versions 1`] = `
"# <meta>
#    <source>https://github.com/PssbleTrngle/Workflows</source>
#    <version>0.0.0-dev</version>
#    <timestamp>2025-01-01T00:00:00.000Z</timestamp>
# </meta>

name: Bug Report
description: File a bug report
labels: ["bug"]

body:
  - type: textarea
    id: what-happened
    attributes:
      label: What happened?
      description: Also tell us, what did you expect to happen?
      placeholder: Tell us what you see!
    validations:
      required: true

  - type: input
    id: version-version
    attributes:
      label: Mod Version
    validations:
      required: true

  - type: input
    id: logs
    attributes:
      label: log output
      description: Please copy and paste your log output to https://mclo.gs/ or a similar site and paste the link here.
"
`;

exports[`minecraft bug templates generates bug template multiple loaders and no versions: bug template multiple loaders and no versions 1`] = `
"# <meta>
#    <source>https://github.com/PssbleTrngle/Workflows</source>
#    <version>0.0.0-dev</version>
#    <timestamp>2025-01-01T00:00:00.000Z</timestamp>
# </meta>

name: Bug Report
description: File a bug report
labels: ["bug"]

body:
  - type: textarea
    id: what-happened
    attributes:
      label: What happened?
      description: Also tell us, what did you expect to happen?
      placeholder: Tell us what you see!
    validations:
      required: true

  - type: dropdown
    id: loader
    attributes:
      label: Loader
      multiple: false
      options:
        - "neoforge"
        - "fabric"
    validations:
      required: true

  - type: input
    id: version-version
    attributes:
      label: Mod Version
    validations:
      required: true

  - type: input
    id: logs
    attributes:
      label: log output
      description: Please copy and paste your log output to https://mclo.gs/ or a similar site and paste the link here.
"
`;

exports[`minecraft bug templates generates bug template multiple loaders and single version: bug template multiple loaders and single version 1`] = `
"# <meta>
#    <source>https://github.com/PssbleTrngle/Workflows</source>
#    <version>0.0.0-dev</version>
#    <timestamp>2025-01-01T00:00:00.000Z</timestamp>
# </meta>

name: Bug Report
description: File a bug report
labels: ["bug"]

body:
  - type: textarea
    id: what-happened
    attributes:
      label: What happened?
      description: Also tell us, what did you expect to happen?
      placeholder: Tell us what you see!
    validations:
      required: true

  - type: dropdown
    id: loader
    attributes:
      label: Loader
      multiple: false
      options:
        - "neoforge"
        - "fabric"
    validations:
      required: true

  - type: input
    id: version-version
    attributes:
      label: Mod Version
    validations:
      required: true

  - type: input
    id: logs
    attributes:
      label: log output
      description: Please copy and paste your log output to https://mclo.gs/ or a similar site and paste the link here.
"
`;

exports[`minecraft bug templates generates bug template multiple loaders and multiple versions: bug template multiple loaders and multiple versions 1`] = `
"# <meta>
#    <source>https://github.com/PssbleTrngle/Workflows</source>
#    <version>0.0.0-dev</version>
#    <timestamp>2025-01-01T00:00:00.000Z</timestamp>
# </meta>

name: Bug Report
description: File a bug report
labels: ["bug"]

body:
  - type: textarea
    id: what-happened
    attributes:
      label: What happened?
      description: Also tell us, what did you expect to happen?
      placeholder: Tell us what you see!
    validations:
      required: true

  - type: dropdown
    id: loader
    attributes:
      label: Loader
      multiple: false
      options:
        - "neoforge"
        - "fabric"
    validations:
      required: true

  - type: input
    id: version-version
    attributes:
      label: Mod Version
    validations:
      required: true

  - type: dropdown
    id: minecraft-version
    attributes:
      label: Minecraft Version
      multiple: false
      options:
        - "1.20.1"
        - "1.21.1"
    validations:
      required: true

  - type: input
    id: logs
    attributes:
      label: log output
      description: Please copy and paste your log output to https://mclo.gs/ or a similar site and paste the link here.
"
`;
